src/java.desktop/macosx/native/libawt_lwawt/java2d/metal/MTLBlitLoops.m

```diff
@@ -800,23 +800,67 @@ void copyFromMTLBuffer(void *pDst, id<MTLBuffer> srcBuf, NSUInteger offset, NSUI
-                    toTexture:dstOps->pTexture destinationSlice:0 destinationLevel:0 destinationOrigin:MTLOriginMake(dstBounds.x1, dstBounds.y1, 0)];
-            [blitEncoder endEncoding];
-
-            [cb commit];
+            struct TxtVertex quadTxVerticesBuffer[6];
+            MTLPooledTextureHandle * interHandle =
+                [mtlc.texturePool getTexture:texWidth
+                                      height:texHeight
+                                      format:MTLPixelFormatBGRA8Unorm];
+            if (interHandle == nil) {
+                J2dTraceLn(J2D_TRACE_ERROR,
+                    "MTLBlitLoops_CopyArea: texture handle is null");
+                return;
+            }
+            [[mtlc getCommandBufferWrapper] registerPooledTexture:interHandle];
+
+            id<MTLTexture> interTexture = interHandle.texture;
+
+            /*
+             * We need to consider common states like clipping while
+             * performing copyArea, thats why we need to query encoder with
+             * appropriate state from EncoderManager and not use
+             * direct MTLBlitCommandEncoder for texture copy.
+             */
+
+            id<MTLRenderCommandEncoder> interEncoder =
+                [mtlc.encoderManager getTextureEncoder:interTexture
+                                           isSrcOpaque:dstOps->isOpaque
+                                           isDstOpaque:dstOps->isOpaque
+            ];
+
+            fillTxQuad(quadTxVerticesBuffer, 0, 0,
+                texWidth, texHeight, texWidth, texHeight, 0,
+                0, texWidth, texHeight, texWidth, texHeight);
+
+            [interEncoder setVertexBytes:quadTxVerticesBuffer
+                             length:sizeof(quadTxVerticesBuffer)
+                            atIndex:MeshVertexBuffer];
+            [interEncoder setFragmentTexture:dstOps->pTexture atIndex: 0];
+            [interEncoder drawPrimitives:MTLPrimitiveTypeTriangle vertexStart:0
+                                                             vertexCount:6];
+
+            id<MTLRenderCommandEncoder> finalEncoder =
+                [mtlc.encoderManager getTextureEncoder:dstOps->pTexture
+                                           isSrcOpaque:dstOps->isOpaque
+                                           isDstOpaque:dstOps->isOpaque
+            ];
+
+            fillTxQuad(quadTxVerticesBuffer, srcBounds.x1, srcBounds.y1,
+                srcBounds.x2, srcBounds.y2, texWidth, texHeight, dstBounds.x1,
+                dstBounds.y1, dstBounds.x2, dstBounds.y2, texWidth, texHeight);
+            [finalEncoder setVertexBytes:quadTxVerticesBuffer
+                             length:sizeof(quadTxVerticesBuffer)
+                            atIndex:MeshVertexBuffer];
+            [finalEncoder setFragmentTexture:interTexture atIndex: 0];
+            [finalEncoder drawPrimitives:MTLPrimitiveTypeTriangle vertexStart:0
```

Comment:
Can't we reuse drawTex2Tex() for this snippet?

Reply:
Thanks for the suggestion Prasanta, yes we should refactor code to use drawTex2Tex. I have update the PR.